### PR TITLE
Add MQTT dispatcher

### PR DESCRIPTION
**Pull request by erwin314**
# Description

This PR add's a MQTT dispatcher. Sending Tesla EV fleet telemetry (e.g., SoC) to an MQTT broker provides real-time, lightweight, and efficient data streaming. MQTT's publish-subscribe model scales well for growing fleets, enabling seamless integration with IoT platforms and data analytics systems for improved monitoring and management.

## Type of change

Please select all options that apply to this change:

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update

# Checklist:

Confirm you have completed the following steps:

- [x] My code follows the style of this project.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding updates to the documentation.
- [x] I have added/updated unit tests to cover my changes.
- [x] I have added/updated integration tests to cover my changes.

### CODE REVIEW

**Comment by iainwhyte** [comment]
OMG, watching intently :)

File: datastore/mqtt/mqtt.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+			publishError = true
+		} else if token.Error() != nil {
+			metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+			p.ReportError("mqtt_dispatch_error", token.Error(), logrus.LogInfo{"topic": "multiple"})
+			publishError = true
+		}
+	}
+
+	// Only process reliable ACK if no errors were reported
+	if !publishError {
```
Comment (agbpatro) [inline datastore/mqtt/mqtt.go:157]: yeah this is tricky with the things you have correctly set up. you are asking the vehicle to resend the whole data if partial uploads to mqtt failed. I think we should just attempt to send the whole payload

File: datastore/mqtt/mqtt.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		metricsRegistry.byteTotal.Add(int64(len(jsonValue)), map[string]string{"record_type": rec.TxType})
+		metricsRegistry.publishCount.Inc(map[string]string{"record_type": rec.TxType})
+	}
+
+	var publishError bool
+	// Wait for all topics to be published
+	for _, token := range tokens {
+		if !token.WaitTimeout(time.Duration(p.config.PublishTimeout) * time.Second) {
+			metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+			p.ReportError("mqtt_publish_timeout", fmt.Errorf("publish operation timed out"), logrus.LogInfo{"topic": "multiple"})
```
Comment (agbpatro) [inline datastore/mqtt/mqtt.go:147]: logging txid and vin will help here in better debugging

File: datastore/mqtt/mqtt.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	}
+}
+
+func (p *MQTTProducer) ReportError(message string, err error, logInfo logrus.LogInfo) {
+	p.airbrakeHandler.ReportLogMessage(logrus.ERROR, message, err, logInfo)
+	p.logger.ErrorLog(message, err, logInfo)
+}
+
+func (p *MQTTProducer) Close() error {
+	p.client.Disconnect(250)
```
Comment (agbpatro) [inline datastore/mqtt/mqtt.go:176]: nit: make `250` a configurable value?

File: datastore/mqtt/mqtt_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+			TopicBase:      "test/topic",
+			QoS:            1,
+			Retained:       false,
+			ConnectTimeout: 30,
+		}
+	})
+
+	Describe("Produce", func() {
+		It("should publish MQTT messages for each field in the payload", func() {
+			// Mock pahomqtt.NewClient
```
Comment (agbpatro) [inline datastore/mqtt/mqtt_test.go:151]: dont think comments in this file are useful

File: datastore/mqtt/mqtt.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	}, nil
+}
+
+func (p *MQTTProducer) Produce(rec *telemetry.Record) {
+	if p.ctx.Err() != nil {
+		return
+	}
+
+	// Alerts and errors are not published to MQTT
+	if rec.TxType == "alerts" || rec.TxType == "errors" {
```
[the remark this answers is not in the document]
Comment (erwin314) [reply datastore/mqtt/mqtt.go:99]: Yes, agreed. I will implement it. Should I convert this PR into a draft?

File: datastore/mqtt/mqtt.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	var tokens []pahomqtt.Token
+
+	// Publish each field to the relevant topic on the MQTT broker
+	for key, value := range convertedPayload {
+		// Skip VIN and CreatedAt fields
+		if key == "Vin" || key == "CreatedAt" {
+			continue
+		}
+
+		topicName := fmt.Sprintf("%s/vin/%s/%s", p.config.TopicBase, rec.Vin, key)
```
[the remark this answers is not in the document]
Comment (erwin314) [reply datastore/mqtt/mqtt.go:127]: No, the MQTT topics have a different structure.

File: datastore/mqtt/mqtt.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+		p.ReportError("mqtt_invalid_payload_type", fmt.Errorf("expected *protos.Payload, got %T", payload), nil)
+		return
+	}
+
+	convertedPayload := transformers.PayloadToMap(data, false, p.logger)
+
+	var tokens []pahomqtt.Token
+
+	// Publish each field to the relevant topic on the MQTT broker
+	for key, value := range convertedPayload {
```
[the remark this answers is not in the document]
Comment (erwin314) [reply datastore/mqtt/mqtt.go:121]: For MQTT topic structuring, it is important to maintain clarity and scalability, especially when dealing with dynamic datasets such as fleet telemetry and event logs.

Sending individual metrics over MQTT rather than combining them in a single topic has several advantages:
- Granular control: Individual topics allow for more precise subscription and filtering. Clients can subscribe only to the specific metrics they need, reducing unnecessary network traffic and processing. A Power Exchange (https://pex.energy) will likely subscript to SoC fields, and a Home Assistant (https://www.home-assistant.io/) might subscribe to location data.
- Easier parsing and debugging: With separate topics, the payload structure for each metric can be simpler and more consistent. This makes parsing and processing easier on the receiving end.
- Scalability: As the system grows, it's easier to add new metrics without modifying existing payload structures or breaking existing subscribers.
- Retention policies: MQTT brokers often allow setting different retention policies per topic, giving you more control over how long each metric's data is stored.

Here is a screenshot from MQTT Explorer demonstrating a clear and scalable topic structure:
Screenshot 2024-10-04 at 08 31 15

File: datastore/mqtt/mqtt.go (hunk 1 above)
Comment (erwin314) [reply datastore/mqtt/mqtt.go:157]: See my [comment above](https://github.com/teslamotors/fleet-telemetry/pull/220#discussion_r1787616172) which explains why the MQTT topic structure is important. Topics are not the same as raw packets from the car.

File: datastore/mqtt/mqtt.go (hunk 2 above)
Comment (erwin314) [reply datastore/mqtt/mqtt.go:147]: I don't think any other part of the existing code does that? And, so far, all txid's I have seen are empty strings. So, could change it, but I doubt if it is needed.

File: datastore/mqtt/mqtt.go (hunk 3 above)
Comment (erwin314) [reply datastore/mqtt/mqtt.go:176]: Yes, that is a good idea. And I'll set the default to 250ms.

File: datastore/mqtt/mqtt_test.go (hunk 4 above)
Comment (erwin314) [reply datastore/mqtt/mqtt_test.go:151]: Ok, will remove them.

File: datastore/mqtt/mqtt.go (hunk 2 above)
Comment (erwin314) [reply datastore/mqtt/mqtt.go:147]: Ok, I had a better look. Agreed, I will implement it.

**Comment by erwin314** [comment]
The code hase been updated with the issues mentioned in the comments above.
- There is now support for MQTT topics: metrics, alerts and errors.
- Other minor changes and fixes.

**Comment by erwin314** [comment]
Add detailed MQTT [README.md](https://github.com/teslamotors/fleet-telemetry/pull/220/commits/b2a58f2d209868811c88d7701f4567ce7b84d09b#diff-8c37119bd842719b2d02ac65a73ac91a2406668123a673ab723cbbdc9e2085b2) explaining architecture, design choices, and configuration

File: config/config.go [hunk 8]
```diff
@@ -312,6 +316,17 @@ func (c *Config) ConfigureProducers(airbrakeHandler *airbrake.AirbrakeHandler, l
 		producers[telemetry.ZMQ] = zmqProducer
 	}
 
+	if _, ok := requiredDispatchers[telemetry.MQTT]; ok {
+		if c.MQTT == nil {
+			return nil, errors.New("Expected MQTT to be configured")
```
[the remark this answers is not in the document]
Comment (erwin314) [reply config/config.go:321]: All other related errors in the same method use a capital letter: 
"Expected ZMQ/Kinesis/PubSub/Kafka/MQTT to be configured"

Maybe it is better if someone creates a separate pull request to fix all error messages? 
(and then maybe also have a look at capital letters used in [errors.py](https://github.com/teslamotors/fleet-telemetry/blob/main/telemetry/errors.go))

File: datastore/mqtt/mqtt.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+		SetPassword(config.Password).
+		SetConnectRetry(true).
+		SetAutoReconnect(true).
+		SetConnectRetryInterval(time.Duration(config.ConnectRetryInterval) * time.Millisecond).
+		SetConnectTimeout(time.Duration(config.ConnectTimeout) * time.Millisecond).
+		SetOrderMatters(false).
+		SetKeepAlive(time.Duration(config.KeepAlive) * time.Second)
+
+	client := PahoNewClient(opts)
+	client.Connect()
```
[the remark this answers is not in the document]
Comment (erwin314) [reply datastore/mqtt/mqtt.go:103]: ConnectRetry is set. So, the client.Connect() might fail, but that's not a problem because the Paho client will keep trying to connect.
This will make sure a telemetry container can start even if a configured MQTT broker is not online.

**Comment by netdata-be** [comment]
@erwin314 is there something I can help with to get this fixed?
I tried to build your code and currently fail to do so.

**Comment by erwin314** [comment]
> @erwin314 is there something I can help with to get this fixed? I tried to build your code and currently fail to do so.

Thanks for reaching out and offering to help! The code was working, but it's been in pull-request mode for a while now. To move forward, perhaps @agbpatro, @aaronpkahn, @ThomasAlxDmy, or @patrickdemers6 could provide feedback on whether this code has a chance of being merged. If there's potential, I’ll update it so it passes all tests again. If merging isn’t feasible, I’m happy to create a dedicated fork focused solely on MQTT.

**Comment by erwin314** [comment]
Hi everyone, I've updated the code, and it should now pass all the tests. However, I'm still unsure if investing my time in this is worthwhile, as it's unclear whether it has a potential for being merged.

**Comment by netdata-be** [comment]
There is 1 suggestion I would like to make, currently you return a json object for each topic, I would rather see the value directly.

Currently:
```
Topic: telemetry/[VIN]/v/Locked
Value: {"value":"true"}
```

Better:
```
Topic: telemetry/VIN/v/Locked
Value: true
```

Why do you put the value inside the key value?
It requires more processing on the consumer to actually read a value.

**Comment by ThomasAlxDmy** [comment]
> > @erwin314 is there something I can help with to get this fixed? I tried to build your code and currently fail to do so.
> 
> Thanks for reaching out and offering to help! The code was working, but it's been in pull-request mode for a while now. To move forward, perhaps @agbpatro, @aaronpkahn, @ThomasAlxDmy, or @patrickdemers6 could provide feedback on whether this code has a chance of being merged. If there's potential, I’ll update it so it passes all tests again. If merging isn’t feasible, I’m happy to create a dedicated fork focused solely on MQTT.

Thank you, this is awesome. We should definitely get it in.

**Comment by erwin314** [comment]
> There is 1 suggestion I would like to make, currently you return a json object for each topic, I would rather see the value directly.
> ...
> Why do you put the value inside the key value? It requires more processing on the consumer to actually read a value.

Thank you for the suggestion and for taking the time to review my pull request! I appreciate your feedback.

If the system evolves to include additional features or richer data, using JSON ensures that existing consumers remain compatible as long as they continue extracting the known keys. In contrast, returning direct values can make maintaining backward compatibility more difficult.

By consistently returning a JSON object, we standardize the data format across all topics, regardless of the type or complexity of the value. This approach allows consumers to parse messages using uniform logic, reducing the likelihood of errors. If some topics return raw values while others return structured data, client-side code must accommodate both cases, leading to increased complexity.

Additionally, JSON maintains the data type of the value, helping to avoid misinterpretation. For example, `{"value": "true"}` clearly indicates a string representation instead of a boolean. This distinction is important because the type of a field may vary by vehicle type or even only by the software version running on it. This might be relevant to the receiving systems.

While it’s true that consumers need to perform an extra step to extract the value, this processing is typically negligible in practice. Parsing JSON is a lightweight operation, supported natively by nearly all programming languages and platforms. On the other hand, returning raw values requires consumers to handle data differently for each topic, which increases the overall complexity of the client-side implementation.

That said, I’m open to further discussion if you feel strongly about this point or have alternative suggestions that might achieve the same benefits with less overhead.

**Comment by erwin314** [comment]
> > > @erwin314 is there something I can help with to get this fixed? I tried to build your code and currently fail to do so.
> > 
> > 
> > Thanks for reaching out and offering to help! The code was working, but it's been in pull-request mode for a while now. To move forward, perhaps @agbpatro, @aaronpkahn, @ThomasAlxDmy, or @patrickdemers6 could provide feedback on whether this code has a chance of being merged. If there's potential, I’ll update it so it passes all tests again. If merging isn’t feasible, I’m happy to create a dedicated fork focused solely on MQTT.
> 
> Thank you, this is awesome. We should definitely get it in.

Thanks for the positive feedback—glad to hear there’s interest in getting this merged! Could you provide some guidance on the steps needed to move this forward?

**Comment by netdata-be** [comment]
Thanks for taking the time to answer.


> Additionally, JSON maintains the data type of the value, helping to avoid misinterpretation. For example, `{"value": "true"}` clearly indicates a string representation instead of a boolean. 

Well, actually that is the problem...

It is a boolean and because you use a transformer from the logger it is casted back to a string, see [here](https://github.com/teslamotors/fleet-telemetry/blob/main/datastore/simple/transformers/payload.go#L63).

And secondly there is actually no need to always include a json with the `value` key.
in the MQTT topic it's always stored as a string, you have to do a json parse anyhow.
And that fixes it is well.

Example in python:

```python
>>> import json
>>> response='{"value": true}'
>>> json.loads(response)
{'value': True}
>>> response='true'
>>> json.loads(response)
True
```

As you can see both are correctly parsed as a boolean.
AS far as I know this is the case in all languages.

So no need to wrap the json in the `value` key.

Just to make sure you see what I mean let's do it again but this time not as a boolean but as a string

```python
>>> import json
>>> response='{"value": "true"}'
>>> json.loads(response)
{'value': 'true'}
>>> response='"true"'
>>> json.loads(response)
'true'
```

So also a string is correctly parsed as a string.

I would suggest to write your json as follows:

```diff
diff --git a/datastore/mqtt/mqtt_payload.go b/datastore/mqtt/mqtt_payload.go
index 3692c3e..595a9b9 100644
--- a/datastore/mqtt/mqtt_payload.go
+++ b/datastore/mqtt/mqtt_payload.go
@@ -19,7 +19,7 @@ func (p *MQTTProducer) processVehicleFields(rec *telemetry.Record, payload *prot
                        continue
                }
                mqttTopicName := fmt.Sprintf("%s/%s/v/%s", p.config.TopicBase, rec.Vin, key)
-               jsonValue, err := json.Marshal(map[string]interface{}{"value": value})
+               jsonValue, err := json.Marshal(value)
                if err != nil {
                        return tokens, fmt.Errorf("failed to marshal JSON for MQTT topic %s: %v", mqttTopicName, err)
                }
```

So to summarize, I think you are using the wrong approach by going to the transformers which seems to be invented for the simple console logging and casting everything to strings.
It would make more sense to use the probuf types when creating the json data, and secondly do not wrap everything a map with the key `value`

**Comment by erwin314** [comment]
> ...
> So to summarize:
> 
> 1. I think you are using the wrong approach by going to the transformers which seems to be invented for the simple console logging and casting everything to strings.
>    It would make more sense to use the probuf types when creating the json data
> 2. Do not wrap everything a map with the key `value` as you could see in the example above, it's not needed to preserve correct json parsing

1. From what I understand, PayloadToMap() doesn’t inherently cast everything to a string. Instead, most fields seem to be using a string-based protobuf defined by the vehicle. However, I'll double-check to confirm. (note: the [link](https://github.com/teslamotors/fleet-telemetry/blob/main/datastore/simple/transformers/payload.go#L63) you mentioned does not seem to show a string cast?)
2. You’ve made a great point about the value being JSON-parsable. This definitely seems like an improvement worth exploring. While we would lose the flexibility to add additional fields, it’s fair to question whether retaining that flexibility might be over-engineering in this case. I'll make the code change and test if it's indeed an improvement a.s.a.p

File: datastore/mqtt/mqtt.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+		return
+	}
+
+	payload := rec.GetProtoMessage()
+
+	var tokens []pahomqtt.Token
+	var err error
+
+	switch payload := payload.(type) {
+	case *protos.Payload:
```
[the remark this answers is not in the document]
Comment (erwin314) [reply datastore/mqtt/mqtt.go:128]: I'll have a look at that

**Comment by erwin314** [comment]
I have updated the way the MQTT values are formatted. They now directly contain the value as json, so they are no-longer wrapped in an extra value entry. And the Invalid field value is now translated to null in json. Even though "null" might not strictly be the same as "invalid", it seems the most straight forward approach.

**Comment by erwin314** [comment]
I've added MQTT support for connectivity events, with one topic per VIN. I assume there's at most one connectivity (dis)connected at a time?

Regarding ConnectionID, it seems to change every time a car (re)connects. This suggests it’s not an ID for a connection to a specific telemetry server, but rather a random ID generated for each new connection. That's why I put it in the value and not in the topic. I'm not sure what happens if a car is connected to multiple telemetry servers (can it even do that?) 

Let me know if there’s any clarification or adjustment needed for this approach!

**Comment by netdata-be** [comment]
@ThomasAlxDmy To me this looks feature complete now. Can you pull some strings to get this unblocked?

**Comment by netdata-be** [comment]
@patrickdemers6 is there something we can do to get this merged?
Is there something missing which needs to be added?

**Comment by andrew-kennedy** [comment]
@erwin314 for your fork, have you published any built docker images? Does your fork run? Debating putting in the effort to run it so I can just get set up with an MQTT dispatcher for telemetry.

**Review by patrickdemers6 — APPROVED** [review]
This is great work, thanks for all your hard work on it!

**Comment by patrickdemers6** [comment]
Looks like there are conflicts to merge. If you can fix those, then let's ship.

**Comment by erwin314** [comment]
> This is great work, thanks for all your hard work on it!

Thank you!

> Looks like there are conflicts to merge. If you can fix those, then let's ship.

Where do you see the merge conflicts? @patrickdemers6
Because Github says: _No conflicts with base branch - Changes can be cleanly merged._

**Comment by patrickdemers6** [comment]
Ah, had to switch to squash and merge instead of rebase.

The best way to learn if this implementation works is to get you and all the others that are interested using it.
